server/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -483,31 +484,38 @@ private Symbol analyzeHaving(Optional<Expression> having,
      *     select name as n  ... order by n
      *     select name  ... order by 1
      *     select name ... order by other_column
+     *     select [1, 2] as arr ... order by arr[1]
      * </pre>
      */
     private static Symbol symbolFromSelectOutputReferenceOrExpression(Expression expression,
                                                                       SelectAnalysis selectAnalysis,
                                                                       String clause,
                                                                       ExpressionAnalyzer expressionAnalyzer,
                                                                       ExpressionAnalysisContext expressionAnalysisContext) {
-        if (expression instanceof QualifiedNameReference) {
-            Symbol symbol = tryGetFromSelectList((QualifiedNameReference) expression, selectAnalysis);
-            if (symbol != null) {
-                return symbol;
+        int ord = -1;
+        if (expression instanceof IntegerLiteral intLiteral) {
+            ord = intLiteral.getValue();
+        } else if (expression instanceof LongLiteral longLiteral) {
+            try {
+                ord = DataTypes.INTEGER.sanitizeValue(longLiteral.getValue());
+            } catch (ClassCastException | IllegalArgumentException e) {
+                throw new IllegalArgumentException(String.format(
+                    Locale.ENGLISH,
+                    "Cannot use %s in %s clause", longLiteral, clause));
             }
         }
-        Symbol symbol = expressionAnalyzer.convert(expression, expressionAnalysisContext);
-        if (symbol instanceof Literal) {
-            symbol = getByPosition(selectAnalysis.outputSymbols(), (Literal<?>) symbol, clause);
+        if (ord > -1) {
+            return ordinalOutputReference(selectAnalysis.outputSymbols(), ord, clause);
         }
+        Symbol symbol = expressionAnalyzer.convert(expression, expressionAnalysisContext);
         return symbol;
     }
```

Comment:
Had to handle the `ORDER BY 1` case differently because the expressionAnalyzer normalizes symbols, so something like `SELECT 123 as foo [...] ORDER BY foo` would return a `123` literal, which then got interpreted as position.